### ISSUE TITLE
Pagination with a "Load more" button

### ISSUE THREAD

**Issue by IvoPereira** [MontFerret/ferret#244]
Hello,

I am looking to a website that has Lazy Loading with a "Load more" button.

What would be the best way to implement this using ferret? I've come across the uncontrolled pagination example - https://github.com/MontFerret/ferret/blob/master/examples/pagination_uncontrolled.fql - but I haven't really been able to reproduce it to this use case.

My example has table tr's, and once the "Load more" button is clicked it keeps appending trs until no more content is available and once so, the button disappears.

A simpler approach (in my point of view) would be to click the button while it exists, and finally, in the end, scrape the entire collection (the initial and the data that was added).

Apart from that, may I ask when are you planning to implement some documentation? It is a little difficult to know some hidden methods that may help in some very specific issues.

If I get to know more about the library I don't mind start helping on it.

Thanks

**Comment by ziflex** [comment]
Hi, thanks for giving Ferret a try!

> but I haven't really been able to reproduce it to this use case.

Why didn't it work? If you take a look [how it works](https://github.com/MontFerret/ferret/blob/master/pkg/stdlib/html/pagination.go#L88), you will notice that it creates an iterator that tries to click a button with a given selector on each iteration. When it cannot find a button, it terminates the iteration.

> Apart from that, may I ask when are you planning to implement some documentation? It is a little difficult to know some hidden methods that may help in some very specific issues.

We are working on that and already have a draft design of the website. So, stay tuned :)

> If I get to know more about the library I don't mind start helping on it.

Thanks, you are very welcome!

**Comment by IvoPereira** [comment]
> Hi, thanks for giving Ferret a try!

The pleasure is mine. This sounds like a great implementation.

 
> Why didn't it work? If you take a look [how it works](https://github.com/MontFerret/ferret/blob/master/pkg/stdlib/html/pagination.go#L88), you will notice that it creates an iterator that tries to click a button with a given selector on each iteration. When it cannot find a button, it terminates the iteration.

The current flow of the website I am trying to scrape is:

1) Show initial data + "Load more" button if available

2) If the element does not exist, scrape the initial data. End.

3) If the element exists, paginate it with the following and keep clicking until the element disappears. (something here is strange, as it looks like it won't stop the PAGINATION loop and keeps doing it (even if in a live example I would only have 3 clicks to be made.

I can make this work for a limited subset if I manually pass a LIMIT 0,3 to get the first 3 batches of added items and wait 3 seconds between them (something in the following lines):

````
LET pageNrs = (
		FOR pageNum IN PAGINATION(doc, loadMoreSelector)
			LIMIT 0,1

			LET loadBtnExists = ELEMENT_EXISTS(doc, loadMoreSelector) ? CLICK(doc, loadMoreSelector) : false
			LET doesExist = loadBtnExists ? WAIT_ELEMENT(doc, loadMoreSelector, 3000) : false

			PRINT("waiting 3 seconds")
			WAIT(3000)

			RETURN {
				pageNum: pageNum
			}
	)
````

One of the issues that might be happening (that's why the wait method) is that a Loading div appears right after clicking the button and only disappears once the content is loaded.

4) Scrape every data available.

TLDR; I am able to get it kinda working but only if limit the results to some values I know beforehand.

> We are working on that and already have a draft design of the website. So, stay tuned :)

Yeah, I've seen that on another issue. It seems pretty cool so far!

> Thanks, you are very welcome!

I will surely will. One of the things that I mainly miss is the lack of an operator list. I find myself skipping through the code to find some operators. Are you looking to get PRs on the current README.md or will you wait for the website docs pages?

Might starting contributing on this.

**Comment by ziflex** [comment]
> The current flow of the website I am trying to scrape is:
> 
> 1. Show initial data + "Load more" button if available
> 2. If the element does not exist, scrape the initial data. End.
> 3. If the element exists, paginate it with the following and keep clicking until the element disappears. (something here is strange, as it looks like it won't stop the PAGINATION loop and keeps doing it (even if in a live example I would only have 3 clicks to be made.

That might be a bug :)

> I can make this work for a limited subset if I manually pass a LIMIT 0,3 to get the first 3 batches of added items and wait 3 seconds between them (something in the following lines):
> 
> ```
> LET pageNrs = (
> 		FOR pageNum IN PAGINATION(doc, loadMoreSelector)
> 			LIMIT 0,1
> 
> 			LET loadBtnExists = ELEMENT_EXISTS(doc, loadMoreSelector) ? CLICK(doc, loadMoreSelector) : false
> 			LET doesExist = loadBtnExists ? WAIT_ELEMENT(doc, loadMoreSelector, 3000) : false
> 
> 			PRINT("waiting 3 seconds")
> 			WAIT(3000)
> 
> 			RETURN {
> 				pageNum: pageNum
> 			}
> 	)
> ```
> One of the issues that might be happening (that's why the wait method) is that a Loading div appears right after clicking the button and only disappears once the content is loaded.

Usually, it's better to wait for an element, instead of plain waiting.  
But in your case, I guess, we need [this function](https://github.com/MontFerret/ferret/issues/229).

> > We are working on that and already have a draft design of the website. So, stay tuned :)
> 
> Yeah, I've seen that on another issue. It seems pretty cool so far!
> 
> > Thanks, you are very welcome!
> 
> I will surely will. One of the things that I mainly miss is the lack of an operator list. I find myself skipping through the code to find some operators. Are you looking to get PRs on the current README.md or will you wait for the website docs pages?
> 
> Might starting contributing on this.

Work on the website will take some time, so let's put information into README first. Later, when we get the fully working website, we can clean it up.

**Comment by zmajew** [comment]
I force through the same 'Load more' issue by this query that simulate 4 cilicks on Load more button.
`
                LET doc = DOCUMENT('https://trends.google.com/trends/trendingsearches/realtime?geo=US&category=all', true)
		LET btn = ELEMENT(doc, '.feed-load-more-button')

		LET res = (
			FOR pageNum IN PAGINATION(doc, '.feed-load-more-button')
			WAIT_ELEMENT(doc, '.feed-load-more-button')
			LIMIT 4
		RETURN pageNum
	)

		WAIT(1000)
		FOR el IN ELEMENTS(doc, '.md-list-block')
			LIMIT 50
			LET ordinal = ELEMENT(el, '.index')
			LET di = ELEMENT(el, '.summary-text')
			LET url = ELEMENT(di, 'a')
			RETURN {
				ordinal: TRIM(ordinal.innerText),
				url: url.attributes.href,
				description: TRIM(url.innerText)
			}
`
Variable 'res' and 'pageNum' does nothing, they are there to make FOR loop to load page in this case 4 times. Hope there is more elegant way to do this.

I like Ferret and the idea. There is examples to start with. Keep going with more examples and documentation: 
I did not find how to approach to the element by its id (<div id="id_name"), and how to insert go variable into query string.

**Comment by ziflex** [comment]
> I force through the same 'Load more' issue by this query that simulate 4 cilicks on Load more button.
```LET doc = DOCUMENT('https://trends.google.com/trends/trendingsearches/realtime?geo=US&category=all', true)
LET btn = ELEMENT(doc, '.feed-load-more-button')

	LET res = (
		FOR pageNum IN PAGINATION(doc, '.feed-load-more-button')
		WAIT_ELEMENT(doc, '.feed-load-more-button')
		LIMIT 4
	RETURN pageNum
)

	WAIT(1000)
	FOR el IN ELEMENTS(doc, '.md-list-block')
		LIMIT 50
		LET ordinal = ELEMENT(el, '.index')
		LET di = ELEMENT(el, '.summary-text')
		LET url = ELEMENT(di, 'a')
		RETURN {
			ordinal: TRIM(ordinal.innerText),
			url: url.attributes.href,
			description: TRIM(url.innerText)
		}
```
> Variable 'res' and 'pageNum' does nothing, they are there to make FOR loop to load page in this case > 4 times. Hope there is more elegant way to do this.

Yes, I guess you need to wait till [this issue](https://github.com/MontFerret/ferret/issues/250) gets resolved (I'm working on it).


> I like Ferret and the idea. There is examples to start with. Keep going with more examples and > documentation:

Glad, that you like it! The documentation will get much better very soon!

> I did not find how to approach to the element by its id (<div id="id_name"), 

You can search by its id using regular CSS selector with the following function ```ELEMENT('#id_name')``` .
Or, you can filter during iteration accessing to element's attributes ```el.attributes.id```.

> and how to insert go variable into query string.

Variables can be passed either via CLI like this:

```
ferret --param=foo:\"bar\", --param=bar:1 ./my-script.fql
```

 or execution context with set variables using one of [these methods](https://github.com/MontFerret/ferret/blob/master/pkg/runtime/options.go#L38) and the variables can be accessed using ```@``` prefix.

**Comment by zmajew** [comment]
Thanks for reply. I assumed that @ prefix is to pull variable into query like $ is in postgres. Can you give example how to pass the variable into query in go script? Something like this one in postgres:
`rows, err := db.Query(`SELECT a	FROM b WHEFE a=$1;`, variableName)`

**Comment by ziflex** [comment]
> Thanks for reply. I assumed that @ prefix is to pull variable into query like $ is in postgres. Can you give example how to pass the variable into query in go script? Something like this one in postgres:
> `rows, err := db.Query(`SELECT a	FROM b WHEFE a=$1;`, variableName)`

I assume, you are using Ferret inside an app, don't you?

Here is a an example, how to pass params:

```
        p := compiler.MustCompile("RETURN @foo")

	out := p.MustRun(
		context.Background(),
		runtime.WithParam("foo", "bar"),
	)
```
